src/Bicep.Core/ArmHelpers/TemplateVariablesEvaluator.cs

```diff
@@ -0,0 +1,336 @@
+        };
+    }
+
+    private bool ReplaceCopyIndexFunctionExpressionsInPlace(FunctionExpression toModify, string expectedCopyIndexName, int loopIteration)
+    {
+        var hasChanges = false;
+
+        if (toModify.Parameters is not null)
+        {
+            for (int i = 0; i < toModify.Parameters.Length; i++)
+            {
+                if (toModify.Parameters[i] is FunctionExpression parameterFunc)
+                {
+                    if (IsCopyIndexFunction(parameterFunc))
+                    {
+                        toModify.Parameters[i] = new JTokenExpression(EvaluateCopyIndexFunction(expectedCopyIndexName,
+                            loopIteration,
+                            parameterFunc.Parameters[0],
+                            parameterFunc.Parameters.Skip(1).FirstOrDefault()));
+                        hasChanges = true;
+                    }
+                    else if (ReplaceCopyIndexFunctionExpressionsInPlace(parameterFunc, expectedCopyIndexName, loopIteration))
+                    {
+                        hasChanges = true;
+                    }
+                }
+            }
+        }
+
+        if (toModify.Properties is not null)
+        {
+            for (int i = 0; i < toModify.Properties.Length; i++)
+            {
+                if (toModify.Properties[i] is FunctionExpression propertyFunc)
+                {
+                    if (IsCopyIndexFunction(propertyFunc))
+                    {
+                        toModify.Properties[i] = new JTokenExpression(EvaluateCopyIndexFunction(expectedCopyIndexName,
+                            loopIteration,
+                            propertyFunc.Parameters[0],
+                            propertyFunc.Parameters.Skip(1).FirstOrDefault()));
+                        hasChanges = true;
+                    }
+                    else if (ReplaceCopyIndexFunctionExpressionsInPlace(propertyFunc, expectedCopyIndexName, loopIteration))
+                    {
+                        hasChanges = true;
+                    }
+                }
+            }
+        }
+
+        return hasChanges;
+    }
+
+    private static bool IsCopyIndexFunction(FunctionExpression expression) => StringComparer.OrdinalIgnoreCase.Equals(expression.Function, CopyIndexFunctionName);
+
+    private static string SerializeExpression(LanguageExpression expression) => expression switch
+    {
+        JTokenExpression { Value: JValue { Value: string strVal } } => $"'{strVal.Replace("'", "''")}'",
+        JTokenExpression otherwise => otherwise.Value.ToString(),
+        FunctionExpression functionExpression => SerializeExpression(functionExpression),
+        _ => throw new InvalidOperationException($"Unrecognized ARM LanguageExpression of type {expression.GetType().Name} encountered."),
+    };
+
+    private static string SerializeExpression(FunctionExpression functionExpression)
```

Comment:
Is there anything in the deployments engine to do this?

Reply:
Not that I could find, although there probably should be. I can open a PR to copy this code into a public method in one of the `Azure.Deployments.` nuget packages, but removing this code from Bicep would need to happen in a future PR.